1.12 Registries Rewrite

LexManos opened the pull request:
The registry system in 1.12 need to be re-written so that it can be more easily reloaded. And so that it supports the ability of overriding things. So that the new Recipe system, will work.

So, this is me re-writing this. I'm opening this as a PR to bring attention to it. And to hopefully allow modders to provide input, and catch the small things that I may miss. If you wish to help out PRs against this branch will be welcome.

So as a overview:
Decoupled vanilla's Registry classes from ours. Modders should NEVER access registries via Vanilla mechanics. Read-only is fine, but it's better to just go through IForgeRegistry.
Removed the concept of substitutions. Things are now just maps, you can register the same name multiple times and the last one will win. Delegates/ObjectHolders will work for all the old ones automatically so no worry there.

Registry events have been moved to directly AFTER PreInit. However modders SHOULD still use these Registry events over Pre/Init/Pos. We have the mechanics of locking the registries down after the registration window is open. So it's best to only register when asked to to prevent issues.
This is done because we need to start introducing the concept of reloading to modders. In the future there may be multiple Registry events fired for the same registry. So this is in preparation.

There are still some tests to run through as this is a MAJOR part of the FML codebase. So I didn't take re-working this lightly. So if anyone wants to help test go for it.

LexManos committed:
First pass of registry rewrite.
Tests do not run yet.
Things of note:
Removed the idea of substitutions, just register multiple items with the same name they will override each other like a normal map.
Decoupled Forge registries from vanilla classes. They now use bouncer classes. MODDERS SHOULD NEVER USE THESE
Introduced more stringent registry locking. As it sits things are only allowed to be registered during the RegistryEvent.Registry phases!
This is to force modders to split up their registrations, and pave the way for calling these functions multiple times during a single MC lifecycle.
ObjectHolder, Missing Mappings, etc... *should* now work for all registry types, nothing should be special cased to Blocks/Items anymore.
Added optional generic dummy factory to registries, to allow registry creators the ability to dummy more then just Blocks.
Re-worked the ItemStack transformer and applied it to ItemBlock and ItemBlockSpecial. Allowing us to cleanup reflective hacks in Block callbacks.
Registry onAdd callback is now only fired on the ACTIVE registry, fixing any issues of temp registries overriding active objects.

LexManos committed:
More work, game loads and runs now.
Moved Registry events to directly AFTER PreInit instead of before.
This allows modders to register handlers for it without @EventBusSubscriber.
It also allows you to register custom things needed before the construction of blocks or items such as Fluids.
TODO: Move Fluids to a real registry.
@ObjectHolder can now be used on private fields.
Reworked FMLModIdMappingEvent to include data for all registries.
Tile Entities are now registrable.

KnightMiner commented:
If registry events are fired after preInit, how will that work for models? As far as I could tell from the code it is still in preInit, but I thought blocks and items had to be registered before you can  register models.

LexManos commented:
There has been a ModelRegistryEvent for quite some time where you should be registering your models. That is fired after all Register events have finished and ObjectHolders updated.

KnightMiner commented:
Also, what about oredict? Traditionally that was done in preInit with items, is it safe do do so during item registry, or does it have it's own event? I assume since the registration may happen multiple times doing so in some init stage is not the best

LexManos commented:
Traditionally, that is wrong. OreDict should be done in either Init or Post init, I would suggest Post init if you're reacting to other people stuff and init if you're working with your own items. OreDict is on my todo list of things to revamp. But for now, thats how you should do it.

Darkhax commented:
This seems like a good change over all. I am very curious to hear your plans for reloading though. You've said in a few other places that recipes won't be reloadable so is there any chance that this will be possible now? Mods like CraftTweaker become much less useful without recipe reloading, and mods like FTBTweaks seem like they just wont be possible without reloading, or being able to modify the recipe system at a later stage.

LexManos commented:
No registry will ever be reloadable while the server is running. This is simply not possible due to the ID syncing. Unless someone goes in and hacks the handshake to resync the ids. Which is out of scope of this change.

thiakil commented:
> Traditionally, that is wrong. OreDict should be done in either Init or Post init

This is what the javadoc says, though https://mcforge.readthedocs.io/en/latest/utilities/oredictionary/ says preInit. As such some modders take this as the one true way, and Tinkers as a specific example requires it in preInit, "because models"

LexManos commented:
Ya, I don't write the documentation. However the way i said it is the way it SHOULD be used. Go make a PR to update the RTDs. Part of the reason that this re-write has mechanics to be so stringent for WHEN you can register things is because nobody agrees on a spot. I've been spouting the right when to do it for years :/

thiakil commented:
Yeah I've certainly been intending to PR a fix as the javadoc is  older than the rtd page, just haven't got around to do it. Was there a specific reason for why it should be init/post? (so I can add more clarity to docs)

KnightMiner commented:
Part of the reason people have accepted pre init is because there is no reason not to register them then. The only dependency of the ore dictionary entries is the blocks or items being registered, so why wait to register them, forcing any compatibility based on the ore dictionary to be delayed?

I assume you are going to say that mods should not be checking which ore dictionary entries are used, but is it really better to manually hardcode support for specific mods adding those items or combinations when it is possible to dynamically add support based on general materials? The end user has much mod compatibility lost as most mods only will add manual support for popular mods anyways.

LexManos commented:
If I understand that complaint correctly you're trying to say that mods are only registering blocks/items if an entry in the OreDict has associated entries. 

First off, You should ALWAYS register your items/blocks. No matter what state the rest of the engine is in. The point of the first pass is to gather a COMPLETE pool of registry objects. So that normal working sets can be created later. Such as connecting to a server with a slightly different modset from you. Or loading a world with a different mod/config set. Ideally there will come a time where you have a 'All the Mods!' mod pack. Everything loads. Multiple versions of each mod load. And then Forge selects the correct content to give you based on what server you connect to. We already have 1/2 the tech to make that work. But it will never work if modders don't register things correctly.

Secondly, OreDict things should ONLY matter when registering recipes, which can have conditions so there is no problem there.

Thirdly, As I said, currently the OreDict does not force a WHEN on the registration of Ores. So any attempt to check if an OreDict entry exists is pointless because it MAY 'exist' after you've checked it. Again, the OreDict is in my TODO list. But outside of the scope of this change. So please get off that subject. I've explained how you SHOULD be using it. Any argument otherwise amounts to "I won't do that because I don't feel like listening to the guy who wrote it!"

@thiakil The basic gist of it is:
Pre-Init: Load configs, do internal things to setup your mod.
Init: Deal with and add your stuff to the game.
Post-Init: React to what others add to the game

liach commented:
Can I send a minor pull that make Entry impl's set registry name look better?

LexManos committed:
Fixup some registry tests, dummies are now persisted across syncing.

HenryLoenwind commented:
Lex, we (Ender IO) have fallback items as subitems on our Material item. So if there's no copper dust, but a copper ingot in the oreDict, we expose it, otherwise it stays hidden. If there's Enderium, we expose a crafting subitem to make it. Same goes for the recipes to make them.

Reason for this is to not pollute JEI/creative menu with useless items, as well as to not register crafting recipes for useless items (e.g. "Enderium Base").

Those things rely on the ore dictionary entries being in place before they are executed. So having a defined stage for that is quite important, because I cannot resolve it if some modder insists on using postInit and some user reports us not seeing that entry as a bug.

I'm not insisting there *must* be a single valid stage enforced by Forge. However, there should be a recommendation, and it would be nice if it was worded something like "or else it's you fault if other mods don't play nice with yours".

LexManos commented:
Subitems mean you still have to register the master item.
Recipes can be conditionally loaded from json with ore info.
As for 'polluting' JEI/Creative, that is not an issue. 
None of this prevents you from registering your master item.

I am still not seeing how having things registered in the oredict has anything to do with whether or not your items exist. Its backward/old thinking there. Your items should always exist. Useless items in the registry is FINE.

This should be the last comment about ore dictionary as it sits. As i'm tired of the circular 'But I MUST know if there is something before I register!' argument. The only answer you will get from me on that is "No you don't, Just register your stuff"

We will not do massive loops over the registration events asking "Are you SURE you don't need to register more stuff, maybe someone else added a ore!". There is no possible way you can have your block based off oredict entries. However you have it now is impossible to work correctly. We can't define a place that entries must be registered.. because they require the Item/Block to be registered and we define that explicitly. But you're arguing that you can't register your items until someone elses items are registered. But they can't register theirs until yours are. This simply does not work and there is no sane way to make it.

HenryLoenwind commented:
Do you really consider other people so stupid that you don't even read what they write and just imagine what they must have written?

diesieben07 commented:
If you don't want stuff to show in CreativeTabs/JEI/etc. don't add anything to the list in `getSubItems`, simple as that.

Xalcon commented:
what about the (memory) overhead of registering blocks, items and models for blocks that will most likely not be used? I never looked into how much memory a block actually needs - and while a single block might not need much, if you have hundreds of blocks, it will add up. Take Chisel as an example. It adds a lot of decorative variations for modded blocks which are (afaik) not registered if the mod is not present. I doubt these entries will be unloaded if they are not needed.

**Personally, I dont mind if I have to register all my blocks all the time - if forge recommends it, I do it.**
The modded community most likely sees this different. Therefore, I still expect mods to allow disabling of item and block registration via its configs, causing possible issues when connecting to servers with a different registry set, just to save a few MB.

/Edit:
Something which just crossed my mind: If we really need to register everything no matter what, what would be the best way of telling MC to not try loading the model+texture for a block? I've a mod that uses a texture from another mod but only registers the block if the mod is present. I dont want forge to vomit a ton of errors just because an optional block couldnt be loaded completly due to some missing (optional!) mod.

LexManos commented:
Alright this is getting off topic. This is about the registry rewrite itself which functionally is identical to the current version. Save for substitutions. That's what we should be discussing. As for the design ideas behind wanting modders to move away from whatever hacks they use during setup. Yes there is a memory cost. Not much even less so if you we define a standard and have people follow it. Any argument around oredicts is nonsensical and outside the scope of this change.

gigaherz commented:
The concept looks fine to me. I tried to go through the code but my brain melted halfway through.

I do however have a doubt/concern: when you speak about overriding things, will there be a possibility for rejecting entries, or only replacing the value registered to one existing resloc with a different instance? That is, I have a mod that pretty much relies on the ability to "remove" some vanilla recipes, so that my alternative mechanics make sense. If only replacement will be possible, I suppose I could make it happen by replacing the vanilla recipes with ones that never accept the inputs, but this seems like a hack, so I wanted to confirm, or at least know that the situation has been considered.

marvin-roesch commented:
Looking good concept-wise as far as I can see, didn't really wade through the code yet, though.

The only minor "concern" I have (which also applies to the current system) is that `IForgeRegistry.Impl`checks for the registry name domain, warning the user it doesn't match the active mod ID. This can already be seen in some mod packs where mods use TCon's API to register metal fluids. If we allow general purpose overrides, that warning seems a little out of place, especially for mods that do nothing but replace Vanilla entries. While I see the purpose of the warning, it suggests to the user that a mod is doing something wrong, in my opinion. If removing it is not an option, I'd at least change the wording to be a little less negative.

LexManos commented:
The situation is being considered and is one of the major reasons for the rewrite, as it sits I've introduced the concept of a mutable registry. But it only has the clear function as of now. Remove is not much of an addition on top of that. However, as this is COMPLETELY new functionality. I WILL be dictating exactly when those functions will be available. Mostly driven by the json system for the recipes. But ya that's not fully done in this commit. As it'll need to be fleshed out when I finish the recipe stuff.

sokratis12GR commented on src/test/java/net/minecraftforge/debug/DynBucketTest.java:
```diff
@@ -94,82 +95,62 @@
-    @Mod.EventBusSubscriber(modid = MODID)
-    public static class Registration
+    @SubscribeEvent
+    public void registrBlocks(RegistryEvent.Register<Block> event)
```
A minor typo

sokratis12GR replied:
A minor typo

sokratis12GR commented on src/test/java/net/minecraftforge/debug/DynBucketTest.java:
```diff
@@ -94,82 +95,62 @@
-            );
-        }
+    @SubscribeEvent
+    public void registrItems(RegistryEvent.Register<Item> event)
```
here too

sokratis12GR commented on src/test/java/net/minecraftforge/debug/DynBucketTest.java:
```diff
@@ -94,82 +95,62 @@
-            GameRegistry.addShapelessRecipe(new ResourceLocation(MODID, "diamond_to_fluid"), null, filledBucket, Ingredient.func_193368_a(Items.DIAMOND));
-        }
+    @SubscribeEvent
+    public void registrRecipes(RegistryEvent.Register<IRecipe> event)
```
here too

AEnterprise commented on src/main/java/net/minecraftforge/common/crafting/CraftingHelper.java:
```diff
@@ -582,20 +584,17 @@ public static void loadRecipes()
-        }
+            reg.clear();
+        else
+            GameData.revert(RegistryManager.FROZEN, GameData.RECIPES, false);
```
won't this cause duplicate recipe registrations after reverting from frozen state? the registry is frozen after all moded recipes are loaded but the loading of those below isn't behind any checks so they would be loaded a 2nd time while already being in the registry

Xalcon replied:
is it actually a problem? Recipes have a unique registry name, they would just replace the existing ones.

AEnterprise replied:
wouldn't be any major issues as far as i see no, it's just that it would mean reading them all from disk again and it with modpacks this could easely be 100-200+ recipe files that would need to be parsed again while the data is already there (unless i'm missing something here)

electrofloat replied:
There are already some modpacks (like FTB Beyond) which need at least 4GB of ram to even start (WTF?) and already taking at least 5 mins to load (loading things multiple times, stiching multiple times, WTF?) because of these "oh, it's fine"/"is it a problem?"/etc.

So no, it's not fine. Yes, it is a problem!

LexManos replied:
Recipes will be loaded from mod jars IDEALLY once. And then frozen like everything else in the registry system. This would gather 99% of recipes.
Then on server start it will load custom recipes from the server/world. This should be a far smaller number. 
Either way the actual recipe loading system should be a fairly trivial amount of that startup time. 
The fact that you say WTF means you haven't done any research into why it does what it does. 
Everything in the MC loading cycle is done for a reason.

AEnterprise replied:
Yes doesnt this cause those mod jar recipes to be loaded again after reverting to the frozen state? Wouldnt it be better to only load if the registry is empty?

LexManos replied:
As it sits this function is only called once and is incomplete. Not calling anything from a modpack or server folder. So honestly it doesn't matter in the slightest. I'll deal with this after merging.

Xalcon commented:
I was wondering, would it make sense to add a registry event for tileentity registation? It feels a bit inconsistent right now.

liach commented:
@Xalcon TEs are using legacy registry for now, so that is not yet supported.

Alpvax commented:
Are there plans to migrate TEs to a ForgeRegistry? When should TEs be registered? preInit? (apologies for being mildly off-topic)

LexManos commented:
Pre, init, post, or Block registration, any should be fine.
There are no plans to move TileEntities to a Forge registry as there is no need to.
They are only ever loaded/saved in their full named format not numerical ids. So expanding them to have all the features of a Forge registry would be wasteful.

KnightMiner commented:
Doesn't the tile entity registry caus a crash if registering the same TE twice? I recall seeing that behavior in 1.7 at least, so if the block registry is intended to reload a couple times in the future registering them with blocks might not be the best place.

Alpvax commented:
Registering them with blocks makes sense, as you cannot have a TileEntity without a block. But I don't see how you could easily accomplish that without either firing a RegisterTileEntityEvent (which would be handled differently to other Registry events) or subclassing RegistryEvent for blocks, to add a registerTileEntity method. Both of which are poor approaches. Improving documentation to include where the recommended place to register them would be.

One approach would be to make the TileEntity registry a different IForgeRegistry (SimpleForgeRegistry for example). That would allow for RegistryEvent<TileEntity> events.

marvin-roesch commented:
What Lex means by "alongside Blocks" is literally *alongside* them, i.e. in the `Register<Block>` event. The problem with making TEs have a fully blown registry (besides being overkill) is that you'd want a mapping `ResourceLocation -> Class<? extends TileEntity>`. Since the registry system is based around `IForgeRegistryEntry` with the registry name methods, you'd either have to wrap that class or do something entirely different. That - imo - isn't worth the effort, simply make it a convention that TEs be registered next to their corresponding blocks, in the block register event.

LexManos commented:
> Doesn't the tile entity registry cause a crash if registering the same TE twice? I recall seeing that behavior in 1.7 at least, so if the block registry is intended to reload a couple times in the future registering them with blocks might not be the best place.

As it sits, No it does not. It's a normal RegistryNamespaced, which means it will LOG a duplicate, but not error. IF this is a case that we need to manage things in the future I would most likely nuke vanilla's registry like I do for the Entity registry and make a different special wrapper registry for them. As Paleo said, currently it's a registry of Class<? extends TileEntity> which is currently impossible to key off of in our system. But luckily as of 1.12 {maybe earlier} the TileEntity registry is a private field, only accessible through TileEntity.register() and TileEntity.getKey() Everything else being internal means we can change it however we want, which includes clearing it and recalling the registry events, without breaking mods. Which is always good.

And yes, when I said block registration, I did mean you could/should register then in the Register<Block> event. Doing it then is logical as it shouldn't need any other dependency, and its intrinsically linked to your block. If you do do it there then if we do decide to refire the Block register event then you'll be golden when it comes to TEs.

But ya, as currently the TE registry is really just a Map<ResourceLocation, Class> and is only ever accessed privately by the TileEntity class by name There are no plans to lift it up to a full managed registry. There are no need for int ids, or anything like that so it shouldn't need all the extra stuff.

So, Just to be clear the OFFICIAL stance of Forge is you should register your TileEntities in the Register<Block> event.

Besides the few typos and policy/timing stuff that have been pointed out there haven't been any real issues brought up. So once I get some time and a bit of git foo, this should be pulled in today/tomorrow. Then i can get back to work on making the Recipe system fully functional for 1.12.

LexManos committed:
Disable substitution test, substitutions are not managed the way they used to be.
We do not care about substitutions, People who replace entries should make them compatible with the existing item.

LexManos commented:
Small hickup, I need to decide how I want to deal with Substitutions. Digging more into cpw's unit tests and he special cases out substitutions. I am trying to decide what is best. Right now we don't care. When a snapshot is injected we just pull whatever is in there. Are there any cases where we should NOT have the substitution? 

Currently we do not teat these special. However I have idea for this. But I would need to see anything people were using this for before. Mainly this may require some form of tagging the 'owner' of the overridden item. So that the server can send a 'I expect you to use minecraft:dirt from mod X' This is most important for things like Recipes. Which if this system is done right in the future would allow for situations where modpacks could 'own' all their recipes. And the same instance of the game could connect to a vanilla server automatically updating it's recipes. Or connect to a modpack server and have those recipes used. But for now this way is fine as it's the status quo.

LexManos committed:
Implement override tracking and syncing.
Overrides MUST be registered within a tracked event so we know what mod it came from.
This will allow servers/saves to select which mod 'wins' and becomes to active entry.
This should also mean that when connecting to a vanilla server things will revert.

LexManos committed:
Fix conflicts

LemADEC commented:
Would that allow to reload recipes on the client when connecting to a server? I think specifically of CraftTweaker allowing a server to add tooltips or fix recipes on the server without updating the client.

gigaherz commented:
I don't think recipe sync is ever going to happen in 1.12 itself, but if I understand correctly, that is going to be a stock feature in 1.13, as part of the datapack stuff. That said, issue/PR comments are not a good place for having such conversations. Please join the forge IRC or Discord if you have more questions like this.
